_posts/18-05-2022-a_story_unfolds.md

```diff
@@ -0,0 +1,223 @@
+```haskell
+buildTree :: File -> IO FileTree
+buildTree file = do
+  isDirectory <- isDir file
+  if isDirectory
+  then do
+    contents <- listDir file
+    subtrees <- traverse buildTree contents
+    return (Directory file subtrees)
+  else return (RegularFile file)
+```
+And we can restore the printing capability by writing another simple function:
+```haskell
+printTree :: FileTree -> String
+printTree = unlines . printTree'
+  where printTree' (RegularFile file) = ['-':(name file)]
+        printTree' (Directory file contents) = ('-':(name file)):(subtrees contents)
+        subtrees = map prepend . concatMap printTree'
+        prepend path = ' ':path
+```
+Reasonably straightforward. We could even get away with making `FileTree` derive `Show` typeclass. The main benefit, though, is that we have separated the process of traversing the filesystem from the process of listing its contents - and should we ever need a function that goes into a directory and removes everything below a certain depth, we already have half of it implemented. Moreover, if we want to recover our initial function, we just need to compose a few blocks:
+```haskell
+tree :: File -> IO ()
+tree file = do
+  dirTree <- buildTree file
+  let formatted = printTree dirTree
+  putStrLn formatted
+```
+But can we go even further?
+
+There seems to be nothing left to extract - we have separated traversal and printing, so what more could we possibly do? But there _is_ in fact a subtle thing. We are walking along the system paths as the trees unfold before us, but we do it in distinct steps: first we take a look around to gather all the possible paths; and then we walk along them.
+
+Let's try to imagine a function like this. First, we need to be able to look around:
+```haskell
+type LookAround a = a -> [a]
+```
+Then, we need to be able to step along the way and construct subtrees - but this is simple recursion. We finally need to be able to construct our node (given the value to store inside the node and the paths to follow):
+```haskell
+type BuildNode t a = a -> [t a] -> t a
+```
+Having these, we can construct our tree builder:
+```haskell
+treeBuilder :: LookAround a -> BuildNode t a -> a -> t a
+treeBuilder lookAround buildNode value = (buildNode value . childNodes . lookAround) value
+  where childNodes = fmap (treeBuilder lookAround buildNode)
+```
+To be more confident that it works as expected, we can, for example, define these:
+```haskell
+data Tree a = Leaf a | Node a [Tree a] deriving Show
+
+children :: LookAround Int
+children 1 = []
+children n = filter (\num -> n `mod` num == 0) [1..n-1]
+
+nodes :: BuildNode Tree Int
+nodes value [] = Leaf value
+nodes value subtrees = Node value subtrees
+```
+
+and then run:
+```haskell
+print (treeBuilder children nodes 12)
+```
+to get a tree in which children of the node are the divisors of the value `v` stored in the node (smaller than `v` itself), `12` being the root.
+
+```
+Node 12 [
```

Comment:
Hmm this is some auto-generated representation? is this `deriving Show` needed for such representation to be available?

Reply:
Yes, this is the representation generated by `deriving Show` - I do however admit to cheating and formatting the output manually.